### PR TITLE
Deprecate dirDelete

### PR DESCRIPTION
**Pull request by scottschurr**
@JoeLoser went to the effort to replace all uses of `dirDelete()` with calls to `ApplyView::dirRemove()`.  This pull request preserves that effort.

There was also an attempt made to deal with the deprecated `dirAdd()` function.  We can't get rid of `dirAdd()` until the `SortedDirectories` amendment has been enabled on the network for at least a year.  But we should go ahead and remove `dirDelete()`.  Don't let the perfect be the enemy of the good.

Reviewers: @mellery451 @wilsonianb

### CODE REVIEW

**Comment by JoeLoser** [comment]
@scottschurr thanks for opening this and reviving the work. Much appreciated!

File: src/ripple/app/tx/impl/CancelCheck.cpp [hunk 1]
```diff
@@ -17,9 +17,10 @@
 */
 //==============================================================================
 
-#include <ripple/app/tx/impl/CancelCheck.h>
 #include <ripple/app/ledger/Ledger.h>
+#include <ripple/app/tx/impl/CancelCheck.h>
```
Comment (wilsonianb) [inline src/ripple/app/tx/impl/CancelCheck.cpp]: Once upon a time, @scottschurr told me:
https://github.com/ripple/validator-keys-tool/pull/1#discussion_r92009139

File: src/ripple/ledger/impl/View.cpp [hunk 2]
```diff
@@ -1216,35 +1021,30 @@ trustDelete (ApplyView& view,
                  beast::Journal j)
 {
     // Detect legacy dirs.
-    bool        bLowNode    = sleRippleState->isFieldPresent (sfLowNode);
-    bool        bHighNode   = sleRippleState->isFieldPresent (sfHighNode);
     std::uint64_t uLowNode    = sleRippleState->getFieldU64 (sfLowNode);
     std::uint64_t uHighNode   = sleRippleState->getFieldU64 (sfHighNode);
     TER         terResult;
```
Comment (wilsonianb) [inline src/ripple/ledger/impl/View.cpp]: It looks like this never gets set. Should we return `tesSUCCESS` at the bottom?

File: src/ripple/ledger/impl/View.cpp [hunk 3]
```diff
@@ -1264,22 +1064,29 @@ offerDelete (ApplyView& view,
...
+    {
         adjustOwnerCount(view, view.peek(
             keylet::account(owner)), -1, j);
+    }
 
     view.erase(sle);
 
-    return (terResult == tesSUCCESS) ?
-        terResult2 : terResult;
+    return tesSUCCESS;
```
Comment (wilsonianb) [inline src/ripple/ledger/impl/View.cpp]: It seems like this should return `tefBAD_LEDGER` if either of the removes weren't successful.

**Comment by JoeLoser** [comment]
@scottschurr I can try to handle them, but I'll happily defer to you when necessary. For example, Mike's question is one that I had myself and still don't know the answer to. Do you or someone with more knowledge about that area of code know, by chance?

File: src/ripple/app/tx/impl/CancelTicket.cpp [hunk 4]
```diff
@@ -80,15 +80,18 @@ CancelTicket::doApply ()
...
-    TER const result = dirDelete (ctx_.view (), false, hint,
-        keylet::ownerDir (ticket_owner), ticketId, false, (hint == 0), viewJ);
+    if (! ctx_.view().dirRemove(
+            keylet::ownerDir(ticket_owner), hint, ticketId, false))
+    {
+        return tefBAD_LEDGER;
+    }
 
+    auto viewJ = ctx_.app.journal ("View");
     adjustOwnerCount(view(), view().peek(
```
[the remark this answers is not in the document]
Comment (scottschurr) [reply src/ripple/app/tx/impl/CancelTicket.cpp]: Good spotting, @mellery451.  This one is messy.  I'll give you my impression of what is afoot here.  We're dealing with a case where a `dirRemove()`, of an object that had better be in the directory, failed.  The failing case _should_ never happen.  If it were to happen it's not clear what the best recovery position would be. And the failing case is hard enough to reproduce that there are no unit tests for it.

So what do we know?

1. If `dirDelete()` had failed in this way, then it would return `tefBAD_LEDGER`.  So we're matching old behavior there.  That's (probably) good.
2. If `doApply()` (the method we're in) returns a code other than `tesSUCCESS` then the transaction is not applied.  So all of the changes we're making here will be rolled back.  (That's _almost_ true -- I'm eliding the `tec`code case since it's not relevant here.)
3. Therefore the new code _should_ accomplish the same ends as the old code but slightly more efficiently.  Since the change to `ownerCount` and the erasure of the `Ticket` will be rolled back because we're returning a `tef` code we don't _need_ to perform those actions.  But there's no harm if we did, since they'll be rolled back.
4. However the extra efficiency is probably irrelevant, since we're hoping this error condition will never occur.  If it does, it means something got fouled up in the ledger.  There should be no way that a _user_ could cause this error to occur.

_To err is human.  To really foul things up requires a computer_ -- Paul Ehrlich

In summary, I _think_ the code is fine as it stands.  And to me it's a hair easier to read.  But it's also fine to switch it back to the previous formulation.

[the remark this answers is not in the document]
Comment (mellery451) [reply src/ripple/app/tx/impl/CancelTicket.cpp]: sounds good - thanks for the explanation.

File: src/ripple/app/tx/impl/CancelCheck.cpp (hunk 1 above)
Comment (JoeLoser) [reply src/ripple/app/tx/impl/CancelCheck.cpp]: Great point. @scottschurr knows best I've quickly come to learn. 👍

File: src/ripple/app/tx/impl/CancelCheck.cpp [hunk 5]
```diff
@@ -104,22 +105,18 @@ CancelCheck::doApply ()
     if (srcId != dstId)
     {
         std::uint64_t const page {(*sleCheck)[sfDestinationNode]};
-        TER const ter {dirDelete (view(), true, page,
-            keylet::ownerDir (dstId), checkId, false, false, viewJ)};
-        if (! isTesSuccess (ter))
+        if (! view().dirRemove (keylet::ownerDir(dstId), page, checkId, false))
```
[the remark this answers is not in the document]
Comment (JoeLoser) [reply src/ripple/app/tx/impl/CancelCheck.cpp]: Good eye. This is a bug I introduced. Thanks!

File: src/ripple/ledger/impl/View.cpp (hunk 2 above)
Comment (JoeLoser) [reply src/ripple/ledger/impl/View.cpp]: Yep, that appears to be the case and I agree with just always returning `tesSUCCESS` at the bottom as we return `tefBAD_LEDGER` if we failed to remove the owner directories.

**Comment by JoeLoser** [comment]
@scottschurr think I've addressed all of the initial feedback. Thanks for the review @wilsonianb and @mellery451.

Feel free to cherry-pick the feedback at [JoeLoser@rippled/54585374b2bb5a7ed7832517a0b1c7ec36d897de](https://github.com/JoeLoser/rippled/commit/54585374b2bb5a7ed7832517a0b1c7ec36d897de)

File: src/ripple/app/tx/impl/CancelCheck.cpp [hunk 6]
```diff
@@ -104,22 +106,18 @@ CancelCheck::doApply ()
...
-            return ter;
+            return tefBAD_LEDGER;
         }
     }
     {
         std::uint64_t const page {(*sleCheck)[sfOwnerNode]};
-        TER const ter {dirDelete (view(), true, page,
-            keylet::ownerDir (srcId), checkId, false, false, viewJ)};
-        if (! isTesSuccess (ter))
+        if (! view().dirRemove (keylet::ownerDir(srcId), page, checkId, false))
```
Comment (wilsonianb) [inline src/ripple/app/tx/impl/CancelCheck.cpp]: This `keepRoot` also got swapped

File: src/ripple/ledger/impl/View.cpp [hunk 7]
```diff
@@ -1216,41 +1021,35 @@ trustDelete (ApplyView& view,
...
+        {
+            return tefBAD_LEDGER;
+        }
     }
 
     JLOG (j.trace()) << "trustDelete: Deleting ripple line: state";
     view.erase(sleRippleState);
 
-    return terResult;
+    return tesSUCCESS;
```
Comment (wilsonianb) [inline src/ripple/ledger/impl/View.cpp]: On second thought, this should return `tefBAD_LEDGER` if the first `dirRemove` fails.

I'm assuming this is a similar situation to what Mike and Scott were discussing
https://github.com/ripple/rippled/pull/2581/files/7e4b9d5a3d51bd083c38c586d219d5219d9c5d04#r196483659
Previously, this would always call `view.erase(sleRippleState)`. Currently, `erase` isn't called if the second `dirRemove` fails. It sounds like it should be fine to immediately return `tefBAD_LEDGER` after each `dirRemove` it fails.

File: src/ripple/app/tx/impl/CancelCheck.cpp (hunk 6 above)
Comment (JoeLoser) [reply src/ripple/app/tx/impl/CancelCheck.cpp]: Good catch, again. I must have been working late when doing this -- apologize for the simple errors. It's pretty evident there are some missing unit tests for these cases though with respect to preserving the original directory as you've pointed out.

File: src/ripple/ledger/impl/View.cpp (hunk 7 above)
Comment (JoeLoser) [reply src/ripple/ledger/impl/View.cpp]: Right, I was thinking about this as well after reading that discussion last night. I agree with your assessment as well. I believe I've implemented what you were suggesting in https://github.com/JoeLoser/rippled/commit/f810d6742dc4139aeec09556bb5bd054ec0ea0da

File: src/ripple/app/tx/impl/CancelCheck.cpp (hunk 6 above)
Comment (JoeLoser) [reply src/ripple/app/tx/impl/CancelCheck.cpp]: Addressed in https://github.com/JoeLoser/rippled/commit/f810d6742dc4139aeec09556bb5bd054ec0ea0da

**Comment by JoeLoser** [comment]
@scottschurr addressed some more of the PR feedback from Brandon today. Feel free to cherry-pick at https://github.com/JoeLoser/rippled/commit/f810d6742dc4139aeec09556bb5bd054ec0ea0da

File: src/ripple/app/tx/impl/SetSignerList.cpp [hunk 8]
```diff
@@ -281,23 +283,23 @@ SetSignerList::removeSignersFromLedger (Keylet const& accountKeylet,
     if (!signers)
         return tesSUCCESS;
 
-    STArray const& actualList = signers->getFieldArray (sfSignerEntries);
+    STArray const& actualList = signers-> getFieldArray (sfSignerEntries);
```
Comment (wilsonianb) [inline src/ripple/app/tx/impl/SetSignerList.cpp]: nit: extra space

File: src/ripple/app/tx/impl/SetSignerList.cpp [hunk 9]
```diff
@@ -281,23 +283,23 @@ SetSignerList::removeSignersFromLedger (Keylet const& accountKeylet,
...
 
-    auto viewJ = ctx_.app.journal ("View");
-    TER const result  = dirDelete(ctx_.view(), false, hint,
-        ownerDirKeylet, signerListKeylet.key, false, (hint == 0), viewJ);
-
-    if (result == tesSUCCESS)
-        adjustOwnerCount(view(),
-            view().peek(accountKeylet), removeFromOwnerCount, viewJ);
+    if (ctx_.view().dirRemove(
+            ownerDirKeylet, hint, signerListKeylet.key, false))
```
Comment (wilsonianb) [inline src/ripple/app/tx/impl/SetSignerList.cpp]: I think this should also return `tefBAD_LEDGER` if `dirRemove` fails.

File: src/ripple/ledger/impl/View.cpp [hunk 10]
```diff
@@ -1264,22 +1065,34 @@ offerDelete (ApplyView& view,
...
 
-    TER terResult  = dirDelete (view, false, sle->getFieldU64 (sfOwnerNode),
-        keylet::ownerDir (owner), offerIndex, false, !bOwnerNode, j);
-    TER terResult2 = dirDelete (view, false, sle->getFieldU64 (sfBookNode),
-        keylet::page (uDirectory), offerIndex, true, false, j);
+    bool const deletedOwner = view.dirRemove(
+        keylet::ownerDir(owner),
+        sle->getFieldU64(sfOwnerNode),
+        offerIndex,
+        false);
```
Comment (wilsonianb) [inline src/ripple/ledger/impl/View.cpp]: This seems fine as is, but I think we could also do what we're doing elsewhere and return `tefBAD_LEDGER` right away if a `dirRemove` fails.

File: src/ripple/app/tx/impl/SetSignerList.cpp (hunk 8 above)
Comment (JoeLoser) [reply src/ripple/app/tx/impl/SetSignerList.cpp]: Not sure how I did that! Just fixed it, though.

File: src/ripple/app/tx/impl/SetSignerList.cpp (hunk 9 above)
Comment (JoeLoser) [reply src/ripple/app/tx/impl/SetSignerList.cpp]: Agreed. Setting consistency with this behavior of always returning `tefBAD_LEDGER` if any of the removals fail will make it easier to reason about.

File: src/ripple/ledger/impl/View.cpp (hunk 10 above)
Comment (JoeLoser) [reply src/ripple/ledger/impl/View.cpp]: Right, I've just made the change so that we just immediately remove `tefBAD_LEDGER` if a `dirRemove` fails. It would be a bit awkward to not be consistent here.

**Comment by JoeLoser** [comment]
Thanks for the great review @wilsonianb. I believe I've addressed your round of comments from tonight in https://github.com/JoeLoser/rippled/commit/e0095bd2a062116f9898d593397906df2a0ec168. Feel free to cherry pick it @scottschurr 👍

**Comment by scottschurr** [comment]
`<soapbox>`
I'd like to use this pull request as an object lesson in the dangers of passing more than one `bool` as a method argument.  The trap was laid by `dirDelete()` which had three distinct `bool` arguments distinguishable only by position.

When `dirDelete()` was converted over to `dirRemove()` (which only has one `bool` in its argument list) it was error prone selecting the _right_ `bool` from `dirDelete()` to transplant into `dirRemove()`.  Two(?) of those errors occurred during the tuning of this pull request.  And those errors were caught visually, not by our unit tests.

If, instead of `bool`, `dirDelete()` had used distinct bool-like types for the three `bool` parameters then the edit would have been a lot less error prone since mistakes would be caught by the compiler.  The bool-like types could have been tagged bools or they could be `enum`s, I don't care.  And if the group agrees that this is a direction we should head, then we should agree on a style.

But I thought, at a minimum, this pull request was an opportunity to point out that we could be using the type system to our advantage more than we have been.
`</soapbox>`

**Comment by nbougalis** [comment]
Merged as 243e181c08db6203f8882bd843f6e858d0d9a25b
